src/AElf.CSharp.CodeOps/Patchers/Module/CallAndBranchCounts/Patcher.cs

```diff
@@ -0,0 +1,94 @@
+using System.Collections.Generic;
+using System.Linq;
+using Mono.Cecil;
+using Mono.Cecil.Cil;
+using Mono.Cecil.Rocks;
+
+namespace AElf.CSharp.CodeOps.Patchers.Module.CallAndBranchCounts;
+
+public class Patcher : IPatcher<ModuleDefinition>
+{
+    public bool SystemContactIgnored => true;
+
+    public void Patch(ModuleDefinition module)
+    {
+        // Check if already injected, do not double inject
+        if (module.Types.Select(t => t.Name).Contains(nameof(ExecutionObserverProxy)))
+            return;
+
+        // ReSharper disable once IdentifierTypo
+        var nmspace = module.Types.Single(m => m.BaseType is TypeDefinition).Namespace;
+
+        var proxyBuilder = new Patch(module, nmspace);
+
+        foreach (var method in module.GetAllTypes().SelectMany(t => t.Methods))
+        {
+            new MethodPatcher(method, proxyBuilder).DoPatch();
+        }
+
+        module.Types.Add(proxyBuilder.ObserverType);
+    }
+}
+
+internal class MethodPatcher
+{
+    private readonly Patch _proxy;
+    private readonly MethodDefinition _method;
+    private bool _alreadyPatched = false;
+    private List<Instruction> _allBranchingInstructions;
+
+    internal MethodPatcher(MethodDefinition method, Patch proxy)
+    {
+        _method = method;
+        _proxy = proxy;
+    }
+
+    private List<Instruction> AllBranchingInstructions
+    {
+        get
+        {
+            if (_allBranchingInstructions == null)
+            {
+                _allBranchingInstructions = _method.Body.Instructions
+                    .Where(i => Constants.JumpingOpCodes.Contains(i.OpCode)).ToList();
+            }
+
+            return _allBranchingInstructions;
+        }
+    }
+
+    public void DoPatch()
+    {
+        if (!_method.HasBody) return;
+        if (_alreadyPatched) return;
+        var processor = _method.Body.GetILProcessor();
+        processor.Body.SimplifyMacros();
+        InsertCallCountAtBeginningOfMethodBody(processor);
+        InsertBranchCountForAllBranches(processor);
+        processor.Body.OptimizeMacros();
+        _alreadyPatched = true;
```

Comment:
Each method will create a new `MethodPatcher`, what does this state `_alreadyPatched` do?

Reply:
The intention was to prevent caller from accidentally calling the method multiple times and screw up the patching.
I think it's just a matter of how we design the contract between the caller and the callee. It's true that currently the caller follows the rule (not calling more than once) implicitly. But there's no harm that we check on the callee side explicitly.